### ISSUE TITLE
Improve `HttpClient` request submission

### ISSUE THREAD

**Issue by NiteshKant** [ReactiveX/RxNetty#282]
In order to have the ability to specify behaviors per request, one has to add these behaviors in `HttpClientConfig`, eg: redirect, user agent, etc. This is odd from the following aspects:
- The ability is hidden in config and hence hard to discover.
- Reuse of requests is difficult as every request has to be created from request if they differ slightly.
### Proposal

Modify `HttpClient.submit()` as:

``` java
public abstract class HttpClient<I, O> {

    public abstract HttpClientRequest<I, O> createGet(String uri);

    public abstract HttpClientRequest<I, O> createPost(String uri);

    public abstract HttpClientRequest<I, O> createPut(String uri);

    public abstract HttpClientRequest<I, O> createDelete(String uri);

    public abstract HttpClientRequest<I, O> createHead(String uri);

    public abstract HttpClientRequest<I, O> createOptions(String uri);

    public abstract HttpClientRequest<I, O> createPatch(String uri);

    public abstract HttpClientRequest<I, O> createTrace(String uri);

    public abstract HttpClientRequest<I, O> createConnect(String uri);

    public abstract HttpClient<I, O> readTimeOut(int timeOut, TimeUnit timeUnit);

    public abstract HttpClient<I, O> followRedirects(int maxRedirects);

    public abstract HttpClient<I, O> followRedirects(boolean follow);

    public abstract <T> HttpClient<I, O> channelOption(ChannelOption<T> option, T value);

    public abstract <II, OO> HttpClient<II, OO> pipelineConfigurator(PipelineConfigurator<OO, II> pipelineConfigurator);

    public abstract HttpClient<I, O> eventLoop(EventLoopGroup eventLoopGroup);

    public abstract HttpClient<I, O> maxConnections(int maxConnections);

    public abstract HttpClient<I, O> withIdleConnectionsTimeoutMillis(long idleConnectionsTimeoutMillis);

    public abstract HttpClient<I, O> withConnectionPoolLimitStrategy(PoolLimitDeterminationStrategy limitDeterminationStrategy);

    public abstract HttpClient<I, O> withPoolIdleCleanupScheduler(ScheduledExecutorService poolIdleCleanupScheduler);

    public abstract HttpClient<I, O> withNoIdleConnectionCleanup();

    public abstract HttpClient<I, O> withNoConnectionPooling();

    public abstract HttpClient<I, O> enableWireLogging(LogLevel wireLogginLevel);

    public abstract HttpClient<I, O> withSslEngineFactory(SSLEngineFactory sslEngineFactory);
}
```

where `HttpClientRequest` is defined as follows:

``` java
public abstract class HttpClientRequest<I, O> extends Observable<HttpClientResponse<O>> {

    protected HttpClientRequest(OnSubscribe<HttpClientResponse<O>> onSubscribe) {
        super(onSubscribe);
    }

    public abstract HttpClientRequest<I, O> withHeader(String name, String value);

    public abstract HttpClientRequest<I, O> withCookie(Cookie cookie);

    public abstract HttpClientRequest<I, O> withContentSource(Observable<I> contentSource);

    public abstract <S> HttpClientRequest<I, O> withRawContentSource(Observable<S> rawContentSource,
                                                                     ContentTransformer<S> transformer);

    public abstract <S> HttpClientRequest<I, O> withRawContent(S content, ContentTransformer<S> transformer);

    public abstract HttpClientRequest<I, O> withContent(I content);

    public abstract HttpClientRequest<I, O> withContent(String content);

    public abstract HttpClientRequest<I, O> withContent(byte[] content);

    public abstract HttpClient<I, O> readTimeOut(int timeOut, TimeUnit timeUnit);

    public abstract HttpClient<I, O> followRedirects(int maxRedirects);

    public abstract HttpClient<I, O> followRedirects(boolean follow);

    public abstract HttpRequestHeaders getHeaders();

    public abstract HttpVersion getHttpVersion();

    public abstract HttpMethod getMethod();

    public abstract String getUri();

    public abstract String getAbsoluteUri();
}
```

Any mutation to this request will create a new request object and hence the created request will be immutable.

Subscription to this request object will execute the request and subscribe the user to the resulting response.

**Comment by benjchristensen** [comment]
Seems to make sense. Would need to use it to know for sure but reviewing the API it seems good.

**Comment by NiteshKant** [comment]
#303 details the API changes discussed here.

**Comment by NiteshKant** [comment]
Available in 0.5.x branch
